### PR TITLE
Add analysis by page rank & by toplevel domain / country

### PR DESCRIPTION
**Pull request by FelixNeutatz**
* not an April 1 joke ;)

### CODE REVIEW

File: analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  public void shadeCountries() {
+    for (Marker marker : countryMarkers) {
+      // Find data for country of the current marker
+      String countryId = marker.getId();
+      DataEntry dataEntry = dataEntriesMap.get(countryId);
+
+      float transparency = 200;
+
+      if (dataEntry != null && dataEntry.value != null) {
+        // Encode value as brightness (values range: 0-1000)
```
Comment (sscdotopen) [inline analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java]: remove commented code

File: analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        float transparency = map(dataEntry.value, 0.0f, 1.0f, 10, 255);
+        System.out.println("transparency: " + transparency);
+        marker.setColor(color(255, 0, 0, transparency));*/
+
+        float red = PApplet.map(dataEntry.value, 0.0f, 1.0f, 0, 255);
+        float blue = PApplet.map((1.0f - dataEntry.value), 0.0f, 1.0f, 0, 255);
+        marker.setColor(color(red, 0, blue, transparency));
+
+
+        } else {
```
Comment (sscdotopen) [inline analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java]: is indentation correct here?

File: analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        // Encode value as brightness (values range: 0-1000)
+        /*
+        float transparency = map(dataEntry.value, 0.0f, 1.0f, 10, 255);
+        System.out.println("transparency: " + transparency);
+        marker.setColor(color(255, 0, 0, transparency));*/
+
+        float red = PApplet.map(dataEntry.value, 0.0f, 1.0f, 0, 255);
+        float blue = PApplet.map((1.0f - dataEntry.value), 0.0f, 1.0f, 0, 255);
+        marker.setColor(color(red, 0, blue, transparency));
+
```
Comment (sscdotopen) [inline analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java]: remove blank lines

File: analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        } else {
+          // No value available
+          marker.setColor(color(100, 120));
+        }
+      }
+    }
+
+    public HashMap<String, DataEntry> loadTrackerDensityFromCSV(String fileName) {
+      HashMap<String, DataEntry> dataEntriesMap = new HashMap<String, DataEntry>();
```
Comment (sscdotopen) [inline analysis/src/main/java/io/ssc/trackthetrackers/analysis/algorithms/plots/ChoroplethMapAppTracker.java]: make this a Map and don't make the type part of the name, just call it dataEntries

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/GraphUtils.scala [hunk 5]
```diff
@@ -26,6 +27,10 @@ object GraphUtils {
     env.readCsvFile[AnnotatedVertex](file, "\n", "\t")
   }
 
+  def readVerticesC(file: String)(implicit env: CollectionEnvironment) = {
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/GraphUtils.scala]: typo in the name?

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala [hunk 6]
```diff
@@ -18,35 +18,46 @@
...
+import scala.io.Source
 
 @deprecated
 object DanglingPageRank extends App {
 
-  pageRank(
-    "/home/ssc/Entwicklung/projects/trackthetrackers/analysis/src/main/resources/cfindergoogle/uris.tsv",
-    "/home/ssc/Entwicklung/projects/trackthetrackers/analysis/src/main/resources/cfindergoogle/links.tsv",
-    15763, .15, 100, 0.0001, "/tmp/flink-scala/pageRanks/")
+  /*
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala]: remove commented code

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala [hunk 7]
```diff
@@ -18,35 +18,46 @@
...
 
-  def pageRank(urisFile: String, linksFile: String, numVertices: Long, teleportationProbability: Double,
+  def pageRank(graphFile: String, domainIndexFile: String, teleportationProbability: Double,
                maxIterations: Int, epsilon: Double, outputPath: String) = {
 
     implicit val env = ExecutionEnvironment.getExecutionEnvironment
 
+    val lines = GraphUtils.readEdges(graphFile)
+
+    val numVertices = lines.map { _ => new Tuple1[Long](1L)}.sum(0)
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala]: space before }

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala [hunk 8]
```diff
@@ -90,20 +102,33 @@ object DanglingPageRank extends App {
...
         (rankFromNeighbors + (danglingRank / numVertices)) * (1.0 - teleportationProbability) +
-        (teleportationProbability / numVertices)
+          (teleportationProbability / numVertices)
 
       RankedVertex(r.id, newRank)
     }
   }
+  
+
+  class InitRanks()
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala]: line break necessary?

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala [hunk 9]
```diff
@@ -90,20 +102,33 @@ object DanglingPageRank extends App {
...
+    extends RichMapFunction[AnnotatedVertex, RankedVertex] {
+
+    override def map(r: AnnotatedVertex): RankedVertex = {
+      RankedVertex(r.id , 1.0 / getRuntimeContext.getBroadcastVariable[Tuple1[Long]]("numVertices").get(0)._1)
+
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DanglingPageRank.scala]: remove blank line

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DegreeDistribution.scala [hunk 10]
```diff
@@ -26,7 +27,7 @@ import org.apache.flink.api.scala._
 @deprecated
 object DegreeDistribution extends App {
 
-  fromEdges("/home/ssc/Entwicklung/projects/trackthetrackers/analysis/src/main/resources/trackinggraph-sample.tsv",
+  fromEdges(Config.get("analysis.trackingraphsample.path"),
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/algorithms/DegreeDistribution.scala]: remove line break

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotCompanyDistribution.scala [hunk 11]
```diff
@@ -47,6 +47,7 @@ object PlotCompanyDistribution extends App {
     dataset.addValue(probability, "", tracker)
   }
 
-  new SingleSeriesBarChart("tracking company distribution", "company", "probability", Color.RED, dataset)
+  new SingleSeriesBarChart("tracking company distribution", "company", "probability", Color.RED, dataset, null)
+  //new SingleSeriesBarChart("tracking company distribution", "company", "probability", Color.RED, dataset, "test.png")
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotCompanyDistribution.scala]: remove commented code

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotCompanyDistributionByRank.scala [hunk 12]
```diff
@@ -0,0 +1,65 @@
+/**
+ * Track the trackers
+ * Copyright (C) 2014  Sebastian Schelter, Felix Neutatz
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotCompanyDistributionByRank.scala]: change to 2015

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotCompanyDistributionByToplevelDomain.scala [hunk 13]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Track the trackers
+ * Copyright (C) 2014  Sebastian Schelter, Felix Neutatz
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotCompanyDistributionByToplevelDomain.scala]: 2015

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotTrackerDistribution.scala [hunk 14]
```diff
@@ -47,6 +47,6 @@ object PlotTrackerDistribution extends App {
     dataset.addValue(probability, "", tracker)
   }
 
-  new SingleSeriesBarChart("tracking domain distribution", "tracking domain", "probability", Color.RED, dataset)
+  new SingleSeriesBarChart("tracking domain distribution", "tracking domain", "probability", Color.RED, dataset, null)
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/PlotTrackerDistribution.scala]: can we make null the default value of the last parameter? than we can avoid these ugly calls

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala [hunk 15]
```diff
@@ -0,0 +1,66 @@
+/**
+ * Track the trackers
+ * Copyright (C) 2014  Sebastian Schelter, Felix Neutatz
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala]: 2015

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+  val lines = Source.fromFile(new File(Config.get("topleveldomainByCountry.csv"))).getLines
+
+  for (domainCountry <- lines) {
+    val splits = domainCountry.split(",")   
+    
+    CompanyDistribution.computeDistribution(Config.get("analysis.trackingraphsample.path"), Config.get("webdatacommons.pldfile.unzipped"),
+      Config.get("analysis.results.path") + "companyDistribution", splits(0).toLowerCase, null, 0) 
+
+    for (file <- new File(Config.get("analysis.results.path") + "companyDistribution").listFiles) {
+      val ll = Source.fromFile(file).getLines
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala]: please give the variable a readable name like lines

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+        if (tokens(0).equals(company)) {
+          writer.write(splits(0) + "," + splits(1) + "," + splits(2) + "," + company + "," + tokens(1) + "\n")
+        }
+      }
+    }
+
+    writer.flush()
+  }
+  writer.close()
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala]: please close the writer in a finally block

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+      }
+    }
+
+    writer.flush()
+  }
+  writer.close()  
+  
+  val mapApp = new ChoroplethMapAppTracker()
+  PApplet.main(Array("io.ssc.trackthetrackers.analysis.algorithms.plots.ChoroplethMapAppTracker"))
+
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/plots/ToplevelCountryMap.scala]: remove blank line

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala [hunk 19]
```diff
@@ -19,29 +19,62 @@
...
+      val pageRankVertices = env.readCsvFile[Rankvertex](pageRankFile, "\n", "\t").map(rank => (rank.vertex, rank.rank))
+
+      val result = pageRankVertices.join(domains)
+        .where(0)
+        .equalTo(0) 
+          { (pageRankVertex, annotatedVertex) => (0, pageRankVertex._1, pageRankVertex._2, annotatedVertex.id)} //url => id
+        .groupBy(0)
+        .sortGroup(2, Order.DESCENDING) //sort by PageRank
+        .first(topKdomains) //filter first k domains
+        //.map { t => (t._4, t._3) } //map to -> (domainID, pageRank)
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala]: remove commented code please

Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala]: remove commented code

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala [hunk 20]
```diff
@@ -19,29 +19,62 @@
...
+        
+      companyEdgesByPageRank = edges.join(result).where("target").equalTo(3)
+                                          { (edge, rankVertex) => edge }
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala]: open brack on top line

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala [hunk 21]
```diff
@@ -19,29 +19,62 @@
...
+      
+    }
+    
+    
+    //filter by domain
+    var companyEdgesByTopleveldomain = companyEdgesByPageRank
+    if ( toplevelDomain != null) {
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala]: no space after (

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala [hunk 22]
```diff
@@ -19,29 +19,62 @@
...
+      val companyDomains =
+        edges.join(domains).where(1).equalTo(1) { (edge, domain) => (edge.src, edge.target, domain.annotation)}
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala]: space before }

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala [hunk 23]
```diff
@@ -19,29 +19,62 @@
...
+
+      companyEdgesByTopleveldomain = companyDomains.filter(domain => domain._3.endsWith(toplevelDomain))
+        .map { tuple => new Edge(tuple._1, tuple._2)}
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala]: space before }

File: analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala [hunk 24]
```diff
@@ -19,29 +19,62 @@
...
+    }
+
+    val numTrackedHosts = companyEdgesByTopleveldomain.distinct("target").map { _ => Tuple1(1L) }.sum(0)
 
-    val companyEdges = edges.filter { edge => Dataset.domainsByCompany.contains(edge.src) }
-                            .map { edge => Dataset.domainsByCompany(edge.src) -> edge.target }
-                            .distinct
+    val companyEdges = companyEdgesByTopleveldomain.filter { edge => Dataset.domainsByCompany.contains(edge.src.toInt) }
```
Comment (sscdotopen) [inline analysis/src/main/scala/io/ssc/trackthetrackers/analysis/statistics/CompanyDistribution.scala]: toInt should be unnecessary

**Comment by FelixNeutatz** [comment]
done :)

**Pull request merged**